File: Computer Programming/C++/Recursion and Dynamic Programming/combinationSum2.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) 
    {
        int size=candidates.size();
        vector<vector<int>> combo;
        vector<int> current;
        sort(candidates.begin(),candidates.end());
        comb(combo,candidates,current,target,0,size);
        return combo;
    }

    void comb(vector<vector<int>> &combo,vector<int>& candidates,vector<int>& current,int target,int index,int size)
    {
        if(target==0)
        {
            combo.push_back(current);
            return;
        }
        if(target<0 || index >= size)
        {
            return;
        }

        current.push_back(candidates[index]);
            comb(combo,candidates,current,target-candidates[index], index+1, size);
        current.pop_back();

        while(index + 1 < size && candidates[index] == candidates[index + 1])
        {
            index++;
        }
        comb(combo,candidates,current,target,index+1,size);
        return;
    }
};

File: Computer Programming/Python/Recursion/Combination Sum 2.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()

        n = len(candidates)
        ans = set()
        current = []

        self.function(ans,current,candidates,target,n-1)
        
        return [list(t) for t in ans]

    def function(self,ans,current,candidates,target,index):
        if target == 0:
            ans.add(tuple(current))
            return

        if target < 0:
            return

        if index < 0:
            return

        current.append(candidates[index])
        self.function(ans,current,candidates,target - candidates[index],index-1)
        current.pop()

        self.function(ans,current,candidates,target,index -1)

        return
